l3backend/l3backend-color.dtx

```diff
@@ -1248,6 +1253,401 @@
+% available as this is pretty burning-edge stuff.
+%    \begin{macrocode}
+if luatexbase.callbacktypes.pre_shipout_filter then
+  luatexbase.add_to_callback(
+    "pre_shipout_filter",
+    function(list)
+      traverse(list,"")
+      return true
+    end,
+    "color")
+end
+%    \end{macrocode}
+%
+% For recent versions of \pkg{luaotfload}, we can register a callback to
+% control how coloring glyph is handled for the color feature.
+%    \begin{macrocode}
+if luaotfload and luaotfload.set_colorhandler then
+  local set_attribute = node.direct.set_attribute
+  luaotfload.set_colorhandler(function(head,n,color)
+    set_attribute(n,attribute,color_index(color))
+    return head,n
+  end)
+end
+%    \end{macrocode}
+%
+% \begin{macro}{luacmd}
+%   Copied from \pkg{l3luatex}: we need to think about how to make this available
+%   (perhaps in |ltx.utils|?).
+%    \begin{macrocode}
+local luacmd do
+  local token_create = token.create
+  local set_lua = token.set_lua
+  local undefined_cs = token.command_id'undefined_cs'
+
+  if not context and not luatexbase then require'ltluatex' end
+  if luatexbase then
+    local new_luafunction = luatexbase.new_luafunction
+    local functions = lua.get_functions_table()
+    function luacmd(name, func, ...)
+      local id
+      local tok = token_create(name)
+      if tok.command == undefined_cs then
+        id = new_luafunction(name)
+        set_lua(name, id, ...)
+      else
+        id = tok.index or tok.mode
+      end
+      functions[id] = func
+    end
+  elseif context then
+    local register = context.functions.register
+    local functions = context.functions.known
+    function luacmd(name, func, ...)
+      local tok = token.create(name)
+      if tok.command == undefined_cs then
+        token.set_lua(name, register(func), ...)
+      else
+        functions[tok.index or tok.mode] = func
+      end
+    end
+  end
+end
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\tex_pdfxform:D}
+%   We redefine the \pkg{expl3} version of the primitive to do the replacement
+%   implicitly. The keywords \texttt{margin} and \texttt{type} are not
+%   documented at the \TeX{} level.
+%    \begin{macrocode}
+luacmd('tex_pdfxform:D', function()
+  local attr = scan_keyword'attr' and scan_string() or nil
+  local resources = scan_keyword'length' and scan_string() or nil
+  local xtype = scan_keyword'type' and scan_string() or nil
+  local margin = scan_keyword'margin' and scan_string() or nil
+  local n = scan_int() or nil
+  traverse(getbox(n),"")
+  saveboxresources(n,attr,resources,false,xtype,margin)
+end,'global','protected')
+%    \end{macrocode}
+% \end{macro}
+%
+%    \begin{macrocode}
+%</lua>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<*package>
+%    \end{macrocode}
+%
+% At present we only set up for \LuaTeX{} in PDF mode: later we will expand
+% on that!
+%
+%    \begin{macrocode}
+%<*luatex>
+%    \end{macrocode}
+%
+% This needs some work: undefine from Lua?
```

Comment:
We can't undefine from Lua (we can use runtoks to run `tex_undefine` while Lua is active, but that's not really from Lua. Might be a useful addition to ltx.util though), but `luacmd` could get a parameter to specify that it is allowed to overwrite. We first have to decide if we really want to overwrite it though.

Reply:
I'll come back to this after we deal with the wider question of what exactly we _do_ want to define.

Reply:
As in, if we are not setting the token `\tex_pdfxform:D`, this is not an issue.

Reply:
I think we are agreed here: we are not touching the primitive name itself, rather a one-level abstraction.